examples/vision/transformer_in_transformer.py

```diff
@@ -0,0 +1,328 @@
+    residual_in_2 = pixel_embedding
+    pixel_embedding = layers.LayerNormalization(epsilon=1e-5)(pixel_embedding)
+    pixel_embedding = mlp(
+        pixel_embedding, in_embedding_dim * mlp_ratio, in_embedding_dim
+    )
+    pixel_embedding = layers.add([pixel_embedding, residual_in_2])
+
+    # outer transformer block
+    _, num_patches, channel = patch_embedding.shape
+    # fusion local and global information
+    fusion_embedding = tf.reshape(
+        pixel_embedding, shape=(-1, num_patches, in_embedding_dim * num_pixel)
+    )
+    fusion_embedding = layers.LayerNormalization(epsilon=1e-5)(fusion_embedding)
+    fusion_embedding = layers.Dense(out_embedding_dim)(fusion_embedding)
+    patch_embedding = layers.add([patch_embedding, fusion_embedding])
+    residual_out_1 = patch_embedding
+    patch_embedding = layers.LayerNormalization(epsilon=1e-5)(patch_embedding)
+    patch_embedding = layers.MultiHeadAttention(
+        num_heads=out_num_heads, key_dim=out_embedding_dim, dropout=attention_dropout
+    )(patch_embedding, patch_embedding)
+    patch_embedding = layers.add([patch_embedding, residual_out_1])
+    residual_out_2 = patch_embedding
+    patch_embedding = layers.LayerNormalization(epsilon=1e-5)(patch_embedding)
+    patch_embedding = mlp(
+        patch_embedding, out_embedding_dim * mlp_ratio, out_embedding_dim
+    )
+    patch_embedding = layers.add([patch_embedding, residual_out_2])
+    return pixel_embedding, patch_embedding
+
+
+"""
+## Implement the TNT model
+The TNT model consists of multiple TNT blocks.
+In the TNT block, there are two transformer blocks where
+the outer transformer block models the global relation among patch embeddings,
+and the inner one extracts local structure information of pixel embeddings.
+The local information is added on the patch
+embedding by linearly projecting the pixel embeddings into the space of patch embedding.
+Patch-level and pixel-level position embeddings are introduced in order to
+retain spatial information.In orginal paper, the authors use the class token for
+classification.
+We use the `layers.GlobalAvgPool1D` to fusion patch information.
+"""
+
+
+def get_model(
+    image_size=96,
+    patch_size=8,
+    outer_block_embedding_dim=64,  # outer transformer block embedding dim
+    inner_block_embedding_dim=16,  # inner transformer block embedding dim
+    num_transformer_layer=5,
+    outer_block_num_heads=4,
+    inner_block_num_heads=2,
+    mlp_ratio=4,
+    attention_dropout=0.2,
+    projection_dropout=0.2,
+    first_stride=4,
+):
+    inputs = layers.Input(shape=input_shape)
+    # Image augment
+    x = data_augmentation(inputs)
+    # extract pixel embedding
+    pixel_embedding, num_patches, new_patch_size = pixel_embed(
+        x, image_size, patch_size, inner_block_embedding_dim, first_stride
+    )
+    num_pixel = new_patch_size[0] * new_patch_size[1]
+    # extract patch embedding
+    patch_embedding = tf.reshape(
```

Comment:
I think this would read clearer if you refactored this part which computes the patch_embedding into it's own patch_embed function (similar to pixel_embed).